subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -756,46 +756,61 @@ private void collectOwnSubgraphConstraints(List<DependencyState> dependencies) {
     }
 
     private void collectAncestorsSubgraphConstraints(List<EdgeState> incomingEdges) {
-        ImmutableSet.Builder<ModuleIdentifier> constraints = null;
+        Set<ModuleIdentifier> constraints = null;
         SubgraphConstraints singleSubgraphConstraints = null;
+        if (incomingEdges.isEmpty()) {
+            ancestorsSubgraphConstraints = SubgraphConstraints.EMPTY;
+            return;
+        }
+        SubgraphConstraints firstParentSubgraphConstraints = incomingEdges.get(0).getFrom().ownSubgraphConstraints;
+        SubgraphConstraints firstParentAncestorsSubgraphConstraints = incomingEdges.get(0).getFrom().ancestorsSubgraphConstraints;
+        if (firstParentSubgraphConstraints.isEmpty() && firstParentAncestorsSubgraphConstraints.isEmpty()) {
+            ancestorsSubgraphConstraints = SubgraphConstraints.EMPTY;
+            return;
```

Comment:
That return is valid because we want the intersection not the union, right?
Might be worth a comment stating that to help future readers of the code.